executor/pipelined_window.go

```diff
@@ -0,0 +1,458 @@
+// Copyright 2021 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package executor
+
+import (
+	"context"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/parser/ast"
+	"github.com/pingcap/tidb/executor/aggfuncs"
+	"github.com/pingcap/tidb/expression"
+	"github.com/pingcap/tidb/planner/core"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/util/chunk"
+)
+
+type dataInfo struct {
+	chk         *chunk.Chunk
+	remaining   uint64
+	accumulated uint64
+}
+
+// PipelinedWindowExec is the executor for window functions.
+type PipelinedWindowExec struct {
+	baseExecutor
+	numWindowFuncs     int
+	windowFuncs        []aggfuncs.AggFunc
+	slidingWindowFuncs []aggfuncs.SlidingWindowAggFunc
+	partialResults     []aggfuncs.PartialResult
+	start              *core.FrameBound
+	end                *core.FrameBound
+	groupChecker       *vecGroupChecker
+
+	// childResult stores the child chunk
+	childResult *chunk.Chunk
+	data        []dataInfo
+	dataIdx     int
+
+	// done indicates the child executor is drained or something unexpected happened.
+	done         bool
+	accumulated  uint64
+	dropped      uint64
```

Comment:
Why we need `dropped` and `dataIdx`? It always processes the first chunk and return it when remaining is 0.

Reply:
No, I originally thought so, but it turns out that even if remaining is 0, e.rows might still references rows in data[0].chk after returned it to upper executor, since there is no guarantee what the upper executor will do to the returned chunk, it might distroy the data (as in the benchmark test, it reused the chunk to pull data, and it will be chk.Reset(), causeing panicing). So I added this check to ensure that chunk will only be returned if there is no row reference.

Reply:
Added comment in the code.